File: crates/iota-sdk/examples/dynamic_fields.rs
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use eyre::Result;
use iota_sdk::graphql_client::Client;

#[tokio::main]
async fn main() -> Result<()> {
    let client = Client::new_localnet();

    // The IOTA system state object owns the validator set and other dynamic
    // fields. It is available on every network including localnet.
    let parent_object_id = "0x5".parse()?;
    let dynamic_fields_page = client
        .dynamic_fields(parent_object_id, Default::default())
        .await?;
    println!("{:#?}", dynamic_fields_page.page_info());
    println!("Page size: {}", dynamic_fields_page.data().len());

    if let Some(first) = dynamic_fields_page.data().first() {
        println!(
            "First field name:\n{}",
            serde_json::to_string_pretty(&first.name)?
        );

        // The field value can be large (e.g. the validator set on 0x5), so we
        // print only the first few lines as a preview.
        let value_pretty = serde_json::to_string_pretty(&first.value_as_json)?;
        const PREVIEW_LINES: usize = 15;
        let mut iter = value_pretty.lines();
        let preview: Vec<&str> = iter.by_ref().take(PREVIEW_LINES).collect();
        let truncated = iter.next().is_some();
        println!("First field value (first {PREVIEW_LINES} lines):");
        println!("{}", preview.join("\n"));
        if truncated {
            println!("... [truncated]");
        }
    }

    Ok(())
}

File: bindings/wasm/examples/dynamic_fields.mjs
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Address, GraphQlClient, initAsync } from "@iota/sdk-wasm";

await initAsync();

const client = GraphQlClient.newTestnet();
const parentObjectId = Address.fromHex(
  "0x7cab491740d51e0d75b26bf9984e49ba2e32a2d0694cabcee605543ed13c7dec",
);

const page = await client.dynamicFields(parentObjectId);
console.log("Page size:", page.data.length);
if (page.data.length > 0) {
  console.log("First field name:\n", page.data[0].name);
  console.log("First field value:\n", page.data[0].valueAsJson);
}
